cpp/modmesh/serialization/SerializableItem.hpp

```diff
@@ -55,65 +56,265 @@ namespace detail
+
+    template <typename T>
+    static void from_json_string(const detail::JsonNodePtr & node, T & value)
     {
-        std::ostringstream oss;
-        oss << "[";
-        const char * separator = "";
-        for (const auto & item : value)
+        if (node->type == detail::JsonType::Null)
+        {
+            return; /* TODO: properly handle null case */
+        }
+
+        if constexpr (detail::is_integer_v<T>)
+        {
+            if (node->type != detail::JsonType::Number)
+            {
+                throw std::runtime_error("Invalid JSON format: invalid number type.");
+            }
+            value = std::stoll(std::get<std::string>(node->value));
+        }
+        else if constexpr (detail::is_float_v<T>)
         {
-            oss << separator << to_json_string(item); /* recursive here */
-            separator = ",";
+            if (node->type != detail::JsonType::Number)
+            {
+                throw std::runtime_error("Invalid JSON format: invalid number type.");
+            }
+            value = std::stod(std::get<std::string>(node->value));
+        }
+        else if constexpr (std::is_same_v<T, std::string>)
+        {
+            if (node->type != detail::JsonType::String)
+            {
+                throw std::runtime_error("Invalid JSON format: invalid number type.");
+            }
+            auto & str = std::get<std::string>(node->value);
+            value = str.substr(1, str.size() - 2); /* Remove quotes */
+        }
+        else if constexpr (std::is_same_v<T, bool>)
+        {
+            if (node->type != detail::JsonType::Boolean)
+            {
+                throw std::runtime_error("Invalid JSON format: invalid boolean type.");
+            }
+            if (std::get<std::string>(node->value) == "false")
+            {
+                value = false;
+            }
+            else
+            {
+                value = true;
+            }
+        }
+        else if constexpr (std::is_same_v<T, std::string>)
+        {
+            if (node->type != detail::JsonType::String)
+            {
+                throw std::runtime_error("Invalid JSON format: invalid number type.");
+            }
+            auto & str = std::get<std::string>(node->value);
+            value = str.substr(1, str.size() - 2); /* Remove quotes */
+        }
+        else if constexpr (std::is_base_of_v<SerializableItem, T>)
+        {
+            if (node->type != detail::JsonType::Object)
+            {
+                throw std::runtime_error("Invalid JSON format: invalid object type.");
+            }
+            value.from_json(std::get<JsonMap>(node->value)); /* recursive here */
+        }
+        else
+        {
+            throw std::runtime_error("Invalid JSON format: invalid type.");
         }
-        oss << "]";
-        return oss.str();
     }
-    else
+
+    template <typename T>
+    static void from_json_string(const JsonNodePtr & node, std::vector<T> & vec)
```

Comment:
This overload can be outside the declaration too.

Reply:
this function is okay to put outside, since it doesn't use the private function. However, as `static void from_json_string(const detail::JsonNodePtr & node, T & value)` must be put in the helper class, I think we should put this function in the helper class for the consistency?

Reply:
By putting outside the (class) declaration, I meant to move the function body, like:

```cpp
class C
{
    void F();
};
void C::F()
{
    // do something
}
```

It has nothing to do with function access.

Reply:
however,  the function is a template function, I don't think we can do it?

Reply:
I see, I realize it's possible to do it even for templates.